src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers.Binary;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+
+namespace System.Reflection.Emit
+{
+    internal readonly struct CustomAttributeWrapper
+    {
+        private readonly ConstructorInfo _constructorInfo;
+        private readonly byte[] _binaryAttribute;
+
+        public CustomAttributeWrapper(ConstructorInfo constructorInfo, ReadOnlySpan<byte> binaryAttribute)
+        {
+            _constructorInfo = constructorInfo;
+            _binaryAttribute = binaryAttribute.ToArray(); // TODO: Update to BlobHandle when public API public APi for MetadataBuilder.GetOrAddBlob(ReadOnlySpan<byte>) added
+        }
+
+        public ConstructorInfo Ctor => _constructorInfo;
+        public byte[] Data => _binaryAttribute;
+    }
+
+    internal struct CustomAttributeInfo
+    {
+        public ConstructorInfo _ctor;
+        public object?[] _ctorArgs;
+        public string[] _namedParamNames;
+        public object?[] _namedParamValues;
+        private const int Field = 0x53;
+        private const int EnumType = 0x55;
+        private const int NullValue = 0xff;
+        private const int OneByteMask = 0x7f;
+        private const int TwoByteMask = 0x3f;
+        private const int FourByteMask = 0x1f;
+
+        internal static CustomAttributeInfo DecodeCustomAttribute(ConstructorInfo ctor, ReadOnlySpan<byte> data)
+        {
+            int pos = 2;
+            CustomAttributeInfo info = default;
+
+            if (data.Length < 2)
+            {
+                throw new InvalidOperationException(SR.Format(SR.InvalidOperation_InvalidCustomAttributeLength, ctor.DeclaringType, data.Length));
```

Comment:
This should be ArgumentException. The argument passed into the method is bad. InvalidOperationException is meant to be used for situations when the object state is invalid - it is not the case here.

The current RefEmit throws ArgumentException as well in this sitution:
```
using System.Reflection;
using System.Reflection.Emit;
using System.Runtime.InteropServices;

AssemblyBuilder ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("MyAssembly"), AssemblyBuilderAccess.Run);
ModuleBuilder mb = ab.DefineDynamicModule("MyModule");
TypeBuilder tb = mb.DefineType("MyType");

tb.SetCustomAttribute(typeof(StructLayoutAttribute).GetConstructors()[0], new byte[] { 42 });
``

Reply:
I have considered throwing `ArgumentException`, but the user could have had used the other overload `tb.SetCustomAttribute(new CustomAttributeBuilder(con, args));` in that case the throwing for byte[] argument would not make sense, so kept the `InvalidOperationException`

Reply:
`CustomAttributeBuilder` should never produce invalid custom attribute blob that would lead to this exception. If this ends up throwing exception for `CustomAttributeBuilder` overload here, it means there is a bug in our code.

Reply:
Good point, updating.